File: src/core/presence_detector.py
from __future__ import annotations

from enum import Enum, auto

try:
    from ai.gaze_tracker import GazeData
except ModuleNotFoundError:
    from ..ai.gaze_tracker import GazeData


class PresenceState(Enum):
    PRESENT_ACTIVE = auto()
    PRESENT_PASSIVE = auto()
    ABSENT = auto()
    UNKNOWN = auto()


class PresenceDetector:
    """Fuse idle time and camera face detection to infer presence."""

    IDLE_THRESHOLD_MS = 300_000
    FACE_ABSENT_FRAMES = 30
    STILLNESS_MOTION_THRESHOLD = 5.0
    STILL_NO_FACE_FRAMES = 20
    DARK_BRIGHTNESS_THRESHOLD = 30.0
    DARK_NO_FACE_FRAMES = 24

    def __init__(self):
        self._face_absent_count = 0
        self._still_no_face_count = 0
        self._dark_no_face_count = 0

    def determine_presence(self, idle_time_ms: int, gaze_data: GazeData | None) -> PresenceState:
        if idle_time_ms < 60_000:
            return PresenceState.PRESENT_ACTIVE

        if gaze_data is None:
            return PresenceState.UNKNOWN

        motion_score = max(0.0, float(getattr(gaze_data, "motion_score", 0.0)))
        brightness = max(0.0, float(getattr(gaze_data, "brightness", 0.0)))

        if not gaze_data.face_detected:
            self._face_absent_count += 1
            if motion_score <= self.STILLNESS_MOTION_THRESHOLD:
                self._still_no_face_count += 1
            else:
                self._still_no_face_count = 0
            if brightness <= self.DARK_BRIGHTNESS_THRESHOLD:
                self._dark_no_face_count += 1
            else:
                self._dark_no_face_count = 0
        else:
            self._face_absent_count = 0
            self._still_no_face_count = 0
            self._dark_no_face_count = 0

        if idle_time_ms >= self.IDLE_THRESHOLD_MS:
            if self._face_absent_count >= self.FACE_ABSENT_FRAMES:
                return PresenceState.ABSENT
            if self._still_no_face_count >= self.STILL_NO_FACE_FRAMES:
                return PresenceState.ABSENT
            if self._dark_no_face_count >= self.DARK_NO_FACE_FRAMES:
                return PresenceState.ABSENT
            if gaze_data.face_detected:
                return PresenceState.PRESENT_PASSIVE

        return PresenceState.UNKNOWN

File: src_cpp/runtime/presence_detector.cpp
#include "runtime/presence_detector.h"

#include <QDateTime>

PresenceDetector::PresenceDetector(PresenceDetectorOptions options)
    : m_options(options)
{
    setTargetFps(m_options.referenceFps);
}

void PresenceDetector::setTargetFps(int targetFps)
{
    m_targetFps = qBound(1, targetFps, 60);
    m_lastSampleMs = -1;
}

PresenceState PresenceDetector::determinePresence(qint64 idleTimeMs, const GazeSample *sample)
{
    if (idleTimeMs < 60000) {
        m_lastSampleMs = -1;
        return PresenceState::PresentActive;
    }

    if (!sample) {
        m_lastSampleMs = -1;
        return PresenceState::Unknown;
    }
    return determinePresence(idleTimeMs, *sample);
}

PresenceState PresenceDetector::determinePresence(qint64 idleTimeMs, const GazeSample &sample)
{
    if (idleTimeMs < 60000) {
        m_lastSampleMs = -1;
        return PresenceState::PresentActive;
    }

    const double elapsedSeconds = sampleSeconds();
    const double motionScore = qMax(0.0, sample.motionScore);
    const double brightness = qMax(0.0, sample.brightness);

    if (!sample.faceDetected) {
        m_faceAbsentSeconds += elapsedSeconds;
        if (motionScore <= m_options.stillnessMotionThreshold) {
            m_stillNoFaceSeconds += elapsedSeconds;
        } else {
            m_stillNoFaceSeconds = 0.0;
        }
        if (brightness <= m_options.darkBrightnessThreshold) {
            m_darkNoFaceSeconds += elapsedSeconds;
        } else {
            m_darkNoFaceSeconds = 0.0;
        }
    } else {
        m_faceAbsentSeconds = 0.0;
        m_stillNoFaceSeconds = 0.0;
        m_darkNoFaceSeconds = 0.0;
    }

    if (idleTimeMs >= m_options.idleThresholdMs) {
        if (m_faceAbsentSeconds >= frameWindowSeconds(m_options.faceAbsentFrames)) {
            return PresenceState::Absent;
        }
        if (m_stillNoFaceSeconds >= frameWindowSeconds(m_options.stillNoFaceFrames)) {
            return PresenceState::Absent;
        }
        if (m_darkNoFaceSeconds >= frameWindowSeconds(m_options.darkNoFaceFrames)) {
            return PresenceState::Absent;
        }
        if (sample.faceDetected) {
            return PresenceState::PresentPassive;
        }
    }

    return PresenceState::Unknown;
}

double PresenceDetector::frameWindowSeconds(int frames) const
{
    return qMax(0.0, static_cast<double>(frames) / static_cast<double>(qMax(1, m_options.referenceFps)));
}

double PresenceDetector::sampleSeconds()
{
    const qint64 nowMs = QDateTime::currentMSecsSinceEpoch();
    const double baselineSeconds = 1.0 / static_cast<double>(qMax(1, m_targetFps));
    if (m_lastSampleMs < 0) {
        m_lastSampleMs = nowMs;
        return baselineSeconds;
    }

    const qint64 elapsedMs = qMax<qint64>(0, nowMs - m_lastSampleMs);
    m_lastSampleMs = nowMs;
    const double elapsedSeconds = static_cast<double>(elapsedMs) / 1000.0;
    return qMax(baselineSeconds, qMin(elapsedSeconds, 1.0));
}
